File: platform/util/ui/src/com/intellij/util/ui/JBImageIcon.java
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.util.ui;

import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import java.awt.*;
import java.awt.image.ImageObserver;

/**
 * HiDPI-aware image icon
 *
 * @author Konstantin Bulenkov
 */
public class JBImageIcon extends ImageIcon {
  public JBImageIcon(@NotNull Image image) {
    super(image);
  }

  @Override
  public synchronized void paintIcon(final Component c, final Graphics g, final int x, final int y) {
    ImageObserver observer = getImageObserver();
    StartupUiUtil.drawImage(g, getImage(), x, y, observer == null ? c : observer);
  }
}

File: platform/util/ui/src/com/intellij/util/ui/JBImageIcon.kt
// Copyright 2000-2025 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.util.ui

import java.awt.Component
import java.awt.Graphics
import java.awt.Image
import javax.swing.ImageIcon

/**
 * HiDPI-aware image icon
 *
 * @author Konstantin Bulenkov
 */
open class JBImageIcon(image: Image) : ImageIcon(image) {
  @Synchronized
  override fun paintIcon(c: Component?, g: Graphics, x: Int, y: Int) {
    StartupUiUtil.drawImage(g, image, x, y, imageObserver ?: c)
  }
}
